examples/examples-log4j/src/main/java/io/grpc/examples/logcontext/CustomLogServer.java

```diff
@@ -0,0 +1,107 @@
+import io.grpc.ServerInterceptors;
+import io.grpc.examples.helloworld.GreeterGrpc;
+import io.grpc.examples.helloworld.HelloReply;
+import io.grpc.examples.helloworld.HelloRequest;
+import io.grpc.stub.StreamObserver;
+import java.io.IOException;
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.message.ReusableMessageFactory;
+import org.apache.logging.log4j.simple.SimpleLogger;
+import org.apache.logging.log4j.util.PropertiesUtil;
+
+/**
+ * A simple server that like {@link io.grpc.examples.helloworld.HelloWorldServer}.
+ * It uses {@link HeaderServerInterceptor} to set the correct logging context for the stub.
+ */
+public class CustomLogServer {
+
+  private static final org.apache.logging.log4j.Logger logger = new SimpleLogger(
+      CustomLogServer.class.getName(),
+      Level.INFO, /* showLogName= */ false,
+      /*showShortLogName=*/ false,
+      /*showDateTime=*/ true,
+      /*showContextMap=*/ true,
+      "yyyy/MM/dd HH:mm:ss:SSS zzz",
+      new ReusableMessageFactory(),
+      PropertiesUtil.getProperties(),
+      System.err);
+
+  /* The port on which the server should run */
+  private static final int PORT = 50051;
+  private Server server;
+
+  private void start() throws IOException {
+    server = ServerBuilder.forPort(PORT)
+        .addService(ServerInterceptors.intercept(new GreeterImpl(), new HeaderServerInterceptor()))
+        .build()
+        .start();
+    logger.info("Server started, listening on " + PORT);
+    Runtime.getRuntime().addShutdownHook(new Thread() {
+      @Override
+      public void run() {
+        // Use stderr here since the logger may have been reset by its JVM shutdown hook.
+        System.err.println("*** shutting down gRPC server since JVM is shutting down");
+        CustomLogServer.this.stop();
+        System.err.println("*** server shut down");
+      }
+    });
+  }
+
+  private void stop() {
+    if (server != null) {
+      server.shutdown();
+    }
+  }
+
+  /**
+   * Await termination on the main thread since the grpc library uses daemon threads.
+   */
+  private void blockUntilShutdown() throws InterruptedException {
+    if (server != null) {
+      server.awaitTermination();
+    }
+  }
+
+  /**
+   * Main launches the server from the command line.
+   */
+  public static void main(String[] args) throws IOException, InterruptedException {
+    final CustomLogServer server = new CustomLogServer();
+    server.start();
+    server.blockUntilShutdown();
+  }
+
+  private static class GreeterImpl extends GreeterGrpc.GreeterImplBase {
+
+    @Override
+    public void sayHello(HelloRequest req, StreamObserver<HelloReply> responseObserver) {
+      logger.info("Got a request");
+      // outputs something like:
+      // 2019/06/05 15:22:12:686 PDT INFO Got a request
+      // {requestId=3e6c256d-6e87-411e-8bf3-fbf81e7ce0e6, clientname=my.domain.name}
+      HelloReply reply = HelloReply.newBuilder().setMessage("Hello " + req.getName()).build();
```

Comment:
Is there any way we can copy the requestId and clientName values into the reply text so on the client side user can see what the serverInterceptor added in the ThreadContext?

Reply:
Also the only one that matters here is the one you set in `onComplete` (or is it `onMessage`) inside the `Listener` in `HeaderServerInterceptor.interceptCall`, right? For the ones set in `onCancel`, `onHalfClose` etc, is there any way to capture and verify them?